base/math.jl

```diff
@@ -103,17 +103,19 @@ function evalpoly(x, p::Tuple)
         _evalpoly(x, p)
     end
 end
+evalpoly(x, ::Tuple{}) = zero(one(x)) # dimensionless zero, i.e. 0 * x^0
 
 evalpoly(x, p::AbstractVector) = _evalpoly(x, p)
 
 function _evalpoly(x, p)
     Base.require_one_based_indexing(p)
     N = length(p)
-    ex = p[end]
+    @inbounds p0 = N == 0 ? reduce_empty_iter(+, p) : p[N]
+    s = oftype(one(x) * p0, p0)
```

Comment:
This construction doesn't seem to guarantee any useful behavior. I think you're assuming that `one(x) isa typeof(x)` is guaranteed for any reasonable `x`? That's not actually part of the interface/contract of `one`, so this doesn't help with type stability in general.

Perhaps something like this would be better (although it'd still cause unnecessary computation when the arithmetic can't be optimized out by the compiler):

```suggestion
    s = oftype(muladd(x, p0, p0), p0)
```

Reply:
`one` seems like it is the correct function here to get the desired behavior with unitful types (where `evalpoly` on an array of coefficients is inherently type-unstable, however).

`x * p0` or `muladd(x, p0, p0)` are both incorrect for unitful `x`.

For non-unitful types, the documentation of `one(x)` says that "If possible, one(x) returns a value of the same type as x".